### DISCUSSION TITLE
Update odoo 15 to odoo 16 running in docker-compose

### DISCUSSION THREAD

**Discussion by mholthausen** [OCA/OpenUpgrade#4120] [General]
When updating from odoo15 to odoo16, two interesting problems occur. First, I don't know if the parameters I set for the upgrade are correct:
```
-c ou_15_16.config --save --database odoo16 --upgrade-path=/mnt/extra-addons/openupgrade_scripts/scripts --load=base,web,account,openupgrade_framework --update all --stop-after-init
```
I get some INFO output and then the upgrade finishes as specified in the parameters. But when I start the image with odoo16, the newly created database is shown as "possibly incompatible", but as compatible when I start version 15 again, as if only a copy of the database is created but no migration is done for the new version. This may also be the reason for this error message:

```
odoo           | 2023-09-03 12:38:35,900 1 ERROR odoo16 odoo.http: Exception during request handling. 
odoo           | Traceback (most recent call last):
odoo           |   File "/usr/lib/python3/dist-packages/odoo/modules/registry.py", line 64, in __new__
odoo           |     return cls.registries[db_name]
odoo           |   File "<decorator-gen-6>", line 2, in __getitem__
odoo           |   File "/usr/lib/python3/dist-packages/odoo/tools/func.py", line 87, in locked
odoo           |     return func(inst, *args, **kwargs)
odoo           |   File "/usr/lib/python3/dist-packages/odoo/tools/lru.py", line 34, in __getitem__
odoo           |     a = self.d[obj]
odoo           | KeyError: 'odoo16'
odoo           | 
odoo           | During handling of the above exception, another exception occurred:
odoo           | 
odoo           | Traceback (most recent call last):
odoo           |   File "/usr/lib/python3/dist-packages/odoo/http.py", line 1998, in __call__
odoo           |     response = request._serve_db()
odoo           |   File "/usr/lib/python3/dist-packages/odoo/http.py", line 1563, in _serve_db
odoo           |     self.registry = Registry(self.db).check_signaling()
odoo           |   File "/usr/lib/python3/dist-packages/odoo/modules/registry.py", line 66, in __new__
odoo           |     return cls.new(db_name)
odoo           |   File "<decorator-gen-14>", line 2, in new
odoo           |   File "/usr/lib/python3/dist-packages/odoo/tools/func.py", line 87, in locked
odoo           |     return func(inst, *args, **kwargs)
odoo           |   File "/usr/lib/python3/dist-packages/odoo/modules/registry.py", line 90, in new
odoo           |     odoo.modules.load_modules(registry, force_demo, status, update_module)
odoo           |   File "/usr/lib/python3/dist-packages/odoo/modules/loading.py", line 510, in load_modules
odoo           |     registry.setup_models(cr)
odoo           |   File "/usr/lib/python3/dist-packages/odoo/modules/registry.py", line 280, in setup_models
odoo           |     env['ir.model']._add_manual_models()
odoo           |   File "/usr/lib/python3/dist-packages/odoo/addons/base/models/ir_model.py", line 446, in _add_manual_models
odoo           |     cr.execute("SELECT *, name->>'en_US' AS name FROM ir_model WHERE state = 'manual'")
odoo           |   File "/usr/lib/python3/dist-packages/odoo/sql_db.py", line 321, in execute
odoo           |     res = self._obj.execute(query, params)
odoo           | psycopg2.errors.UndefinedFunction: operator does not exist: character varying ->> unknown
odoo           | LINE 1: SELECT *, name->>'en_US' AS name FROM ir_model WHERE state =...
odoo           |                       ^
odoo           | HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.
```

The second point is that the tables of the installed invoicing app "account_*" are not transferred and completely missing in the new database, although it is specified in the load parameter.

Do you have any advice on what I may be missing here?

**Comment by legalsylvain** [comment]
Why do you load account module ?

**Reply by mholthausen** [reply]
Thanks for your reply. The database tables of the `account` module were not migrated and so I thought it was missing from this listing of modules. But it makes no difference if I put it there or not.
But that is one of my problems, that account is not transferred to the new database at all.

**Reply by mholthausen** [reply]
I think I found out what's wrong. 
At first I created a new database, so all previously installed modules were missing. When I upgraded the existing database, the respective tables stayed in the database. 
Second, I adjusted the parameters into `--database odoo --upgrade-path=/mnt/extra-addons/openupgrade_scripts/scripts --load=base,web,openupgrade_framework --update all --stop-after-init`, started the migration and then used the `odoo:16.0` image to start it again. 

It looks like that some migrations did not run successfully or need some further action, especially on the `ir_*` tables (jsonb fields, see the error):
```
odoo           | psycopg2.errors.UndefinedFunction: operator does not exist: character varying ->> unknown
odoo           | LINE 1: SELECT *, name->>'en_US' AS name FROM ir_model WHERE state =...
```

Is there something I have to execute after the general migration?

**Reply by mholthausen** [reply]
After I manually adjusted the responding columns into jsonb data type I encounter more problems. It seems some other tables aren't migrated completely:
- `ERROR: column res_company.active does not exist`
- `ERROR: column res_company.message_main_attachment_id does not exist`
- `psycopg2.errors.InFailedSqlTransaction: current transaction is aborted, commands ignored until end of transaction block`
- and the warning `odoo odoo.addons.base.models.ir_cron: Skipping database odoo as its base version is not 16.0.1.3.` what I thought the migration would do.

Is it at all already possible to upgrade from odoo 15 to odoo 16 with OpenUpgrade?
